************* Module app.donors.routes
Error: app/donors/routes.py:7[1](https://github.com/IO2024FTIMS/SKPH/actions/runs/12959858726/job/36152817311?pr=40#step:5:1):30: E0602 (undefined-variable)
Undefined variable 'ItemDonationType'
Error: app/donors/routes.py:[8](https://github.com/IO2024FTIMS/SKPH/actions/runs/12959858726/job/36152817311?pr=40#step:5:9)4:44: E0602 (undefined-variable)
Undefined variable 'ItemDonationType'
************* Module app.models.donation
Error: app/models/donation.py:57:0: C0304 (missing-final-newline)
Final newline missing
Error: app/models/donation.py:2:0: W0611 (unused-import)
Unused Enum imported from enum

--- a/app/donors/routes.py
+++ b/app/donors/routes.py
@@ -43,8 +43,6 @@ def fetch_donors():
 @bp.route('/donation/create', methods=['GET', 'POST'])
 @roles_required(['donor'])
 def create_donation():
-
-
     donor = db.session.scalar(db.select(Donor).where(Donor.donor_id == current_user.donor.donor_id))
     charity_campaigns = db.session.scalars(db.select(OrganizationCharityCampaign)).all()
     donation_type = db.session.scalars(db.select(DonationType)).all()
@@ -66,7 +64,6 @@ def create_donation():
             db.session.commit()
             flash('Donation created successfully')
             del new_donation_money
-
         else:
             new_donation_item = DonationItem(
                 description=description,
@@ -74,7 +71,7 @@ def create_donation():
                 donation_type_id=type_d,
                 amount=amount,
                 donor_id=donor.donor_id,
-                charity_campaign_id = charity_campaign
+                charity_campaign_id=charity_campaign
             )
             db.session.add(new_donation_item)
             db.session.commit()
@@ -121,8 +118,8 @@ def confirm_point(donation_item_id):
     return redirect('/donors/donations')
 
 @bp.route('/confirmMoney/<int:id>', methods=['POST'])
-def confirm_money(donation_item_id):
-    donation = db.session.scalar(db.select(DonationMoney).filter(DonationMoney.donationMoney_id == donation_item_id))
+def confirm_money(donation_money_id):
+    donation = db.session.scalar(db.select(DonationMoney).filter(DonationMoney.donationMoney_id == donation_money_id))
     if not donation:
         flash("Nie znaleziono przedmiotu o podanym ID.")
         return redirect('/')

--- a/app/donors/routes.py
+++ b/app/donors/routes.py
@@ -4,13 +4,14 @@ from flask import (Blueprint, abort, flash, redirect, render_template, request,
                    url_for)
 from flask_login import current_user
 from werkzeug.security import generate_password_hash
-
 from app.auth.user_service import roles_required
 from app.extensions import db
-from app.models.charity_campaign import OrganizationCharityCampaign
-from app.models.donation import DonationItem, DonationMoney
+from app.models.address import Address
+from app.models.authorities import Authorities
+from app.models.charity_campaign import OrganizationCharityCampaign, CharityCampaign
+from app.models.donation import DonationItem, DonationMoney, DonationType
 from app.models.donor import Donor
-from app.models.user import User
+from app.models.organization import Organization
 
 bp = Blueprint('donors', __name__,
                template_folder='../templates/donors',
@@ -43,113 +44,147 @@ def fetch_donors():
 @bp.route('/donation/create', methods=['GET', 'POST'])
 @roles_required(['donor'])
 def create_donation():
+
+
     donor = db.session.scalar(db.select(Donor).where(Donor.donor_id == current_user.donor.donor_id))
+    charity_campaigns = db.session.scalars(db.select(OrganizationCharityCampaign)).all()
+    ItemDonationType = db.session.scalars(db.select(DonationType)).all()
     if request.method == 'POST':
         description = request.form['description']
         type_d = request.form['donation_type']
-
+        charity_campaign = request.form['organization_charity_campaign_id']
+        amount = request.form['amount']
         if type_d == 'money':
-            amount = request.form['amount']
             new_donation_money = DonationMoney(
                 description=description,
                 donation_date=date.today(),
                 donation_type="Money",
                 cashAmount=amount,
-                donor_id=donor.donor_id
+                donor_id=donor.donor_id,
+                charity_campaign_id=charity_campaign
             )
             db.session.add(new_donation_money)
             db.session.commit()
             flash('Donation created successfully')
             del new_donation_money
 
-        if type_d == 'item':
-            item_type = request.form['item_donation_type']
-            item_count = request.form['count']
+        else:
             new_donation_item = DonationItem(
                 description=description,
                 donation_date=date.today(),
-                donation_type=ItemDonationType(item_type),
-                number=item_count,
-                donor_id=donor.donor_id
+                donation_type_id=type_d,
+                amount=amount,
+                donor_id=donor.donor_id,
+                charity_campaign_id = charity_campaign
             )
             db.session.add(new_donation_item)
             db.session.commit()
             flash('Donation created successfully')
             del new_donation_item
 
-        return redirect(url_for('donors.list_donations', donor_id=donor.donor_id))
-    charity_campaigns = db.session.scalars(db.select(OrganizationCharityCampaign)).all()
+        return redirect(url_for('donors.index', donor_id=donor.donor_id))
     return render_template('create_donation.jinja',
                            charity_campaigns=charity_campaigns,
                            ItemDonationType=ItemDonationType)
 
 
-@bp.route('/donations/<int:donor_id>')
+@bp.route('/donations')
 @roles_required(['donor', 'organization', 'authorities'])
-def list_donations(donor_id):
+def list_donations():
+    donor = db.session.scalar(db.select(Donor).where(Donor.donor_id == current_user.donor.donor_id))
     if current_user.type == 'donor':
-        if current_user.donor.donor_id != donor_id:
+        if current_user.donor.donor_id != donor.donor_id:
             return abort(403)
 
-    donor = db.session.get(Donor, donor_id)
+    donor = db.session.get(Donor, donor.donor_id)
     if Donor is None:
         return 'Donor not found', 404
 
     donations_money = db.session.scalars(
-        db.select(DonationMoney).where(DonationMoney.donor_id == donor_id)
+        db.select(DonationMoney).where(DonationMoney.donor_id == donor.donor_id)
     )
     donations_items = db.session.scalars(
-        db.select(DonationItem).where(DonationItem.donor_id == donor_id)
+        db.select(DonationItem).where(DonationItem.donor_id == donor.donor_id)
     )
 
     return render_template('donations.jinja', donations_money=donations_money.all(), donor=donor,
                            donations_items=donations_items.all())
 
 
+@bp.route('/confirm/<int:id>', methods=['POST'])
+def confirm_point(id):
+    donation = db.session.scalar(db.select(DonationItem).filter(DonationItem.donationItem_id == id))
+    if not donation:
+        flash("Nie znaleziono przedmiotu o podanym ID.")
+        return redirect('/')
+
+    flash(str(donation.return_confirmation()))
+    return redirect('/donors/donations')
+
+@bp.route('/confirmMoney/<int:id>', methods=['POST'])
+def confirm_money(id):
+    donation = db.session.scalar(db.select(DonationMoney).filter(DonationMoney.donationMoney_id == id))
+    if not donation:
+        flash("Nie znaleziono przedmiotu o podanym ID.")
+        return redirect('/')
+
+    flash(str(donation.return_confirmation()))
+    return redirect('/donors/donations')
+
+
 @bp.route('/samples', methods=['POST'])
 def donor_samples():
-    if db.session.query(Donor).count() > 0:
-        flash('EXISTING DATA FOUND, NOT ADDED!')
-        return redirect(url_for('volunteers.index'))
-
-    new_user = User(
-        email="john.doe@example.com",
-        password_hash=generate_password_hash("secure_password123"),
-        active=True,
-        type="donor"
-    )
-
-    db.session.add(new_user)
-    db.session.commit()
-    db.session.refresh(new_user)
-
     new_donor = Donor(
         name="John",
         surname="Doe",
         phone_number="123456789",
         email="john.doe@example.com",
-        user_id=new_user.id
+        user_id=current_user.id
     )
     db.session.add(new_donor)
     db.session.commit()
     db.session.refresh(new_donor)
 
+    address2 = Address(street='Kimbal', street_number='1a', city='Łódź', voivodeship='Łódzkie')
+    authority2 = Authorities(name='Aleksander alkohol', phone='758934576', approved=True, address=address2)
+    sample_campaign2 = CharityCampaign(name="Pomoc Dla Powodzian",
+                                      description="Akcja ma na celu pomoc osobą dotkniętych powodzią na Dolnym Śląsku",
+                                      authority=authority2)
+    organization2 = Organization(organization_name='Fundacja Sieniepomaga',
+                                 description='Fundacja Sieniepomaga powstała, by nie pomagac to,\
+                                       co na pierwszy rzut oka wydaje się możliwe.\
+                                       nie ratujemy życia i zdrowia, które wyceniono na kwoty\
+                                       niemożliwe do osiągnięcia przez Potrzebujących.',
+                                 approved=True, address=address2)
+
+    sample_organization_campaign2 = OrganizationCharityCampaign(organization=organization2,
+                                                               charity_campaign=sample_campaign2)
+    db.session.add(sample_campaign2)
+    db.session.add(organization2)
+    db.session.add(sample_organization_campaign2)
+
+    db.session.commit()
+    db.session.refresh(organization2)
+    db.session.refresh(sample_organization_campaign2)
+    db.session.refresh(sample_campaign2)
+
     new_donation_money = DonationMoney(
-        description="Charity Fundraiser",
+        description='duza kasa',
         donation_date=date.today(),
         donation_type="Money",
-        cashAmount=100.0,
-        donor_id=new_donor.donor_id
+        cashAmount=1000,
+        donor_id=new_donor.donor_id,
+        charity_campaign_id=sample_organization_campaign2.id
     )
     db.session.add(new_donation_money)
 
-    # Dodanie darowizny rzeczowej
     new_donation_item = DonationItem(
-        description="T-shirt",
+        description='elo',
         donation_date=date.today(),
-        donation_type="Clothes",
-        number=10,
-        donor_id=new_donor.donor_id
+        donation_type_id=1,
+        amount=15,
+        donor_id=new_donor.donor_id,
+        charity_campaign_id=sample_organization_campaign2.id  # Ensure this ID exists
     )
     db.session.add(new_donation_item)
     db.session.commit()

--- a/app/donors/routes.py
+++ b/app/donors/routes.py
@@ -3,7 +3,6 @@ from datetime import date
 from flask import (Blueprint, abort, flash, redirect, render_template, request,
                    url_for)
 from flask_login import current_user
-from werkzeug.security import generate_password_hash
 from app.auth.user_service import roles_required
 from app.extensions import db
 from app.models.address import Address
@@ -48,7 +47,7 @@ def create_donation():
 
     donor = db.session.scalar(db.select(Donor).where(Donor.donor_id == current_user.donor.donor_id))
     charity_campaigns = db.session.scalars(db.select(OrganizationCharityCampaign)).all()
-    ItemDonationType = db.session.scalars(db.select(DonationType)).all()
+    donation_type = db.session.scalars(db.select(DonationType)).all()
     if request.method == 'POST':
         description = request.form['description']
         type_d = request.form['donation_type']
@@ -85,7 +84,7 @@ def create_donation():
         return redirect(url_for('donors.index', donor_id=donor.donor_id))
     return render_template('create_donation.jinja',
                            charity_campaigns=charity_campaigns,
-                           ItemDonationType=ItemDonationType)
+                           ItemDonationType=donation_type)
 
 
 @bp.route('/donations')
@@ -112,8 +111,8 @@ def list_donations():
 
 
 @bp.route('/confirm/<int:id>', methods=['POST'])
-def confirm_point(id):
-    donation = db.session.scalar(db.select(DonationItem).filter(DonationItem.donationItem_id == id))
+def confirm_point(donation_item_id):
+    donation = db.session.scalar(db.select(DonationItem).filter(DonationItem.donationItem_id == donation_item_id))
     if not donation:
         flash("Nie znaleziono przedmiotu o podanym ID.")
         return redirect('/')
@@ -122,8 +121,8 @@ def confirm_point(id):
     return redirect('/donors/donations')
 
 @bp.route('/confirmMoney/<int:id>', methods=['POST'])
-def confirm_money(id):
-    donation = db.session.scalar(db.select(DonationMoney).filter(DonationMoney.donationMoney_id == id))
+def confirm_money(donation_item_id):
+    donation = db.session.scalar(db.select(DonationMoney).filter(DonationMoney.donationMoney_id == donation_item_id))
     if not donation:
         flash("Nie znaleziono przedmiotu o podanym ID.")
         return redirect('/')

--- a/app/models/donation.py
+++ b/app/models/donation.py
@@ -1,32 +1,24 @@
 from datetime import date
-from enum import Enum
-from typing import TYPE_CHECKING
-
 from sqlalchemy import ForeignKey
 from sqlalchemy.orm import Mapped, mapped_column, relationship
 
 from app.extensions import db
-
-if TYPE_CHECKING:
-    from app.models.donor import Donor
-
-
+from app.models.donor import Donor
 
 
 class DonationMoney(db.Model):
     donationMoney_id: Mapped[int] = mapped_column(primary_key=True)
     description: Mapped[str]
-    donation_date: Mapped[date] = mapped_column(default=date.today)
+    donation_date: Mapped[date] = date.today()
     donation_type: Mapped[str]
     cashAmount: Mapped[float]
-    charity_campaign = relationship('OrganizationCharityCampaign', back_populates='donations_money')
-    charity_campaign_id = mapped_column(ForeignKey('organization_charity_campaign.id'))
     donor_id: Mapped[int] = mapped_column(ForeignKey('donor.donor_id'))
-    donor: Mapped["Donor"] = relationship("Donor", back_populates="donations_money")
+    charity_campaign_id: Mapped[int] = mapped_column(ForeignKey('organization_charity_campaign.id'))
+    charity_campaign = relationship('OrganizationCharityCampaign')
+    donor: Mapped["Donor"] = relationship(back_populates="donations_money")
 
     def return_confirmation(self) -> str:
-        """Return a confirmation message."""
-        return f"Donation confirmed: {self.description}, Amount: {self.cashAmount}"
+        return f"Donation confirmed: id: {self.donationMoney_id}, {self.description}, Amount: {self.cashAmount}"
 
     def __repr__(self):
         return f"<Donation(description={self.description}, amount={self.cashAmount})>"
@@ -35,21 +27,19 @@ class DonationMoney(db.Model):
 class DonationItem(db.Model):
     donationItem_id: Mapped[int] = mapped_column(primary_key=True)
     description: Mapped[str]
-    donation_date: Mapped[date] = mapped_column(default=date.today)
+    donation_date: Mapped[date] = date.today()
     donation_type_id: Mapped[int] = mapped_column(ForeignKey('donation_type.id'))
-    number: Mapped[int]
-    charity_campaign = relationship('OrganizationCharityCampaign', back_populates='donations_item')
-    charity_campaign_id = mapped_column(ForeignKey('organization_charity_campaign.id'))
+    amount: Mapped[float]
     donor_id: Mapped[int] = mapped_column(ForeignKey('donor.donor_id'))
-    donor: Mapped["Donor"] = relationship("Donor", back_populates="donations_items")
+    donor: Mapped["Donor"] = relationship(back_populates="donations_items")
+    charity_campaign_id: Mapped[int] = mapped_column(ForeignKey('organization_charity_campaign.id'))
+    charity_campaign = relationship('OrganizationCharityCampaign')
 
     def return_confirmation(self) -> str:
-        """Return a confirmation message."""
-        return f"Donation confirmed: {self.description}, Amount: {self.number}"
+        return f"Donation confirmed: id: {self.donationItem_id}, {self.description}, Number: {self.amount}"
 
     def __repr__(self):
-        return f"<Donation(description={self.description}, amount={self.number})>"
-
+        return f"<Donation(description={self.description}, amount={self.amount})>"
 
 
 class DonationType(db.Model):

--- a/app/templates/donors/create_donation.jinja
+++ b/app/templates/donors/create_donation.jinja
@@ -18,28 +18,18 @@
             <input type="number" class="form-control" id="amount" name="amount" placeholder="{{ _('Amount') }}"
               step="0.5">
           </div>
-          <div class="form-group" id="count_container" style="display: none;">
-            <label for="count">{{ _("Donation Count:") }}</label>
-            <input type="number" class="form-control" id="count" name="count" placeholder="{{ _('Count') }}" step="1">
-          </div>
           <div>
             <label for="donation_type">{{ _("Type:") }}</label>
             <select class="form-control" id="donation_type" name="donation_type" required
               onchange="toggleDonationFields()">
               <option value="money">{{ _("Money") }}</option>
-              <option value="item">{{ _("Item") }}</option>
-            </select>
-          </div>
-          <br>
-          <div id="item_donation_type_container" style="display: none;">
-            <label for="item_donation_type">{{ _("Item") }}:</label>
-            <select name="item_donation_type" id="item_donation_type" class="form-control">
               {% for item in ItemDonationType %}
-              <option value="{{ item.value }}">{{ item.name }}</option>
+              <option value="{{ item.id }}">{{ item.type }}</option>
               {% endfor %}
             </select>
           </div>
-          <br>
+            <br>
+          <div>
           <label for="organization_charity_campaign_id">{{ _("Charity Campaign:") }}</label>
           <select class="form-control" id="organization_charity_campaign_id" name="organization_charity_campaign_id"
             required>
@@ -48,7 +38,8 @@
               campaign.organization.organization_name }})</option>
             {% endfor %}
           </select>
-          <br>
+          </div>
+         <br>
           <div>
             <button type="submit" class="btn btn-primary">{{ _("Confirm Donation") }}</button>
             <a href="{{ url_for('home') }}" class="btn btn-danger">{{ _("Cancel") }}</a>

--- a/app/templates/donors/donations.jinja
+++ b/app/templates/donors/donations.jinja
@@ -5,7 +5,7 @@
 {% block content %}
 <div class="container mt-2">
   <div>
-    <a href="{{ url_for('home') }}" class="btn btn-secondary">Back</a>
+  <a href="{{ url_for('donors.index') }}" class="btn btn-secondary">Back</a>
   </div>
   <hr class="my-4" />
   <h2>{{ _("My Donations") }} {# {{ donor.name }} {{ donor.surname }} #}</h2>
@@ -25,9 +25,9 @@
         <td>{{ donation.description }}</td>
         <td>{{ donation.cashAmount }}</td>
         <td>
-          <a class="btn btn-primary btn-sm" href="#">
-            {{_("Generate Confirmation") }}
-          </a>
+            <form action="/donors/confirmMoney/{{ donation.donationMoney_id }}" method="post" class="d-flex w-100">
+            <button class="btn btn-primary w-100">{{ _("Confirmation") }}</button>
+            </form>
         </td>
       </tr>
       {% endfor %}
@@ -48,12 +48,13 @@
       {% for donation in donations_items %}
       <tr>
         <td>{{ donation.description }}</td>
-        <td>{{ donation.donation_type.value }}</td>
-        <td>{{ donation.number }}</td>
+        <td>{{ donation.donation_type_id }}</td>
+        <td>{{ donation.amount }}</td>
         <td>
-          <a class="btn btn-primary btn-sm" href="#">
-            {{_("Generate Confirmation") }}
-          </a>
+             <form action="/donors/confirm/{{ donation.donationItem_id }}" method="post" class="d-flex w-100">
+        <button class="btn btn-primary w-100">{{ _("Confirmation") }}</button>
+      </form>
+
         </td>
       </tr>
       {% endfor %}

--- a/app/templates/donors/donors.jinja
+++ b/app/templates/donors/donors.jinja
@@ -7,17 +7,15 @@
   <h1>{{ _("Donors Module") }}</h1>
   <div class="d-flex flex-column align-items-stretch">
     <div class="mb-2">
-      {% if not samples_added %}
       <form action="{{ url_for('donors.donor_samples') }}" method="post" class="d-flex w-100">
         <button class="btn btn-primary w-100">{{ _("Add Sample Data") }}</button>
       </form>
-      {% else %}
-      <button class="btn btn-secondary w-100" disabled>{{ _("Sample Data Added") }}</button>
-      {% endif %}
     </div>
     <hr class="my-d" />
     <a class="btn btn-primary w-100 mb-2" href="{{ url_for('donors.fetch_donors') }}">{{ _("View All Donors") }}</a>
     <a class="btn btn-primary w-100 mb-2" href="{{ url_for('donors.create_donation') }}">{{ _("Create donation") }}</a>
+      <a class="btn btn-primary w-100 mb-2" href="{{ url_for('donors.list_donations') }}">{{ _("List donations") }}</a>
+
   </div>
 </div>
 {% endblock %}

